runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -895,11 +897,20 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
    TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
    TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
    TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
-   TR::Block *indirectAccessBlock;
+
+   //if arrayBlockNeeded is false, we haven't generated and we don't need arrayDirectAccessBlock
+   TR::Block *arrayDirectAccessBlock = arrayBlockNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
    TR::Block * directAccessBlock;
-   if (conversionNeeded)
+   TR::Block *indirectAccessBlock;
+
+   // There are 5 possible cases determining which blocks should be added to the generated IL trees
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded: add directAccess, indirectAccess, and arrayDirectAccess
+   // 2.) checksNeeded && arrayCheckNeeded && !arrayBlockNeeded: add directAccess and indirectAccess only
```

Comment:
which scenario is case2? some comments are due here. i meant: if arrayBlockNeeded is false, how can arrayCheckNeeded be true? what is the arrayChecking for? i.e. if it is an array (the checking result), you still don't need arrayBlock. why bother to check it to begin with?

Reply:
I believe this is a holdover from before we decided to check if the object class is known at compile time (i.e.: all type checks were performed at runtime no matter what). This case would only occur with when using gencon GC policy, if the J9Class of the object could not be retrieved (i.e.: `javaLangClass == NULL`). But you're right that the array check shouldn't be necessary in the first place since the only condition that changes whether a direct/indirect access should be used is if the object type is `java/lang/Class`.